src/rust/en.omegascans/src/parser.rs

```diff
@@ -0,0 +1,226 @@
+		cover,
+		title,
+		author,
+		artist,
+		description,
+		url,
+		status: manga_status,
+		nsfw: MangaContentRating::Nsfw,
+		viewer: MangaViewer::Scroll,
+		..Default::default()
+	})
+}
+
+pub fn parse_chapter_list(base_url: String, manga_id: String) -> Result<Vec<Chapter>> {
+	let url = format!("{}/series/{}", BASE_API_URL, manga_id);
+	let data = Request::new(url, HttpMethod::Get).json()?.as_object()?;
+
+	let mut all_chapters: Vec<Chapter> = Vec::new();
+
+	let seasons = data.get("seasons").as_array()?;
+
+	for season in seasons {
+		let season = season.as_object()?;
+		let chapters = season.get("chapters").as_array()?;
+
+		for chapter in chapters {
+			let chapter = chapter.as_object()?;
+			let price = chapter.get("price").as_int()?;
+
+			// Only get free chapters
+			if price != 0 {
+				continue;
+			}
+
+			let index = chapter.get("index").as_string()?.read();
+			let id = chapter.get("chapter_slug").as_string()?.read();
+			let url = format!("{}/series/{}/{}", base_url, manga_id, id);
+
+			let date_updated = chapter
+				.get("created_at")
+				.as_date("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Some("en_US"), None)
+				.unwrap_or(-1.0);
+
+			all_chapters.push(Chapter {
+				id,
+				chapter: index.parse::<f32>().unwrap_or_default(),
+				date_updated,
+				url,
+				..Default::default()
+			});
+		}
+	}
+
+	Ok(all_chapters)
+}
+
+pub fn parse_page_list(base_url: String, manga_id: String, chapter_id: String) -> Result<Vec<Page>> {
+	let url = format!("{}/series/{}/{}", base_url, manga_id, chapter_id);
+
+	let obj = Request::new(url, HttpMethod::Get).html()?;
+
+	let mut page_list: Vec<Page> = Vec::new();
+
+	for (i, page) in obj.select("img").array().enumerate() {
+		let obj = page.as_node().expect("node array");
+		let mut url = obj.attr("data-src").read();
+
+		if url.is_empty() {
+			url = obj.attr("src").read();
+		}
+
+		page_list.push(Page {
+			index: i as i32,
+			url,
+			..Default::default()
+		});
+	}
+
+	// Remove icon.png and banners from top and bottom
+	page_list.remove(0);
+	page_list.remove(0);
+	page_list.pop();
+	page_list.pop();
+
+	Ok(page_list)
+}
+
+pub fn modify_image_request(base_url: String, request: Request) {
+	request.header("Referer", &base_url);
+}
+
+fn parse_manga(base_url: &String, json: Request) -> Result<Vec<Manga>> {
+	let data = json.json()?.as_object()?.get("data").as_array()?;
+	let mut mangas: Vec<Manga> = Vec::new();
+
+	for manga in data {
+		let manga = manga.as_object()?;
+		let title = manga.get("title").as_string()?.read();
+		let cover = manga.get("thumbnail").as_string()?.read();
+		let id = manga.get("series_slug").as_string()?.read();
+
+		let url = format!("{}/series/{}", base_url, id);
+
+		mangas.push(Manga {
+			id,
+			cover,
+			title,
+			url,
+			..Default::default()
+		});
+	}
+
+	Ok(mangas)
+}
+
+fn is_last_page(url: &String) -> bool {
+	let json = Request::new(url, HttpMethod::Get);
+	!json.json().expect("").as_object().expect("").get("data").as_array().expect("").is_empty()
```

Comment:
- you're re-requesting the page even when you already have a json from it (probably just check if `manga` is empty)
- don't use `expect` unless you are certain it will never be called. this will break the source until the app is restarted
- make sure there's a newline at the end of the file

Reply:
Oops, sorry about that. Forgot I could just check the manga array. I used expect since it said I couldn't use ? in a bool function, but that's unnecessary anyway. Hope this is the last fix, thanks for the code reviews.